### PR TITLE
Legend: maxWidth and maxLinesPerEntry

### PR DESCRIPTION
**Pull request by CalvinFernandez**
Adds `maxWidth` property to legend
 - When defined, maxWidth restricts legend growth to at most `maxWidth` pixels.

Adds `maxLinesPerEntry` property to legend
- When defined, `maxLinesPerEntry` specifies number of times a single entry may wrap before being truncated

Refactors legend to include `LegendTable` to aid with layout calculation.

### CODE REVIEW

**Comment by CalvinFernandez** [comment]
closes #3161 #3162

**Comment by jchomicki** [comment]
@CalvinFernandez What is the status on this? A downstream ticket is blocking on this.

**Comment by CalvinFernandez** [comment]
@jchomicki waiting on reviewers. ping @hellochar @themadcreator

File: src/components/legend.ts
Comment (themadcreator) [inline src/components/legend.ts:530]: This isn't really the idiomatic way to do d3. Removing and re-adding everything on every render causes lots of DOM thrashing. You can see from the code that this replaces that it applies the rows to the DOM using the `.data` on the selection then using `.enter`, `.exit` and the selection itself to update the stuff that's there.

File: src/components/legend.ts
Comment (themadcreator) [inline src/components/legend.ts]: Why the extra `Infinity`?

File: src/components/legend.ts
Comment (themadcreator) [inline src/components/legend.ts]: even/odd index detection for symbol/text determination seems pretty fragile

File: src/components/legend.ts
Comment (themadcreator) [inline src/components/legend.ts]: don't declare this function every call. Instead make it a private static method on `Legend` or add it to `Utils`

File: src/components/legend.ts
Comment (themadcreator) [inline src/components/legend.ts:478]: is `_calculateLayoutInfo` the conventional name for this in other components? If so, we should still use it. If not, whatever.

File: src/components/legend.ts
Comment (themadcreator) [inline src/components/legend.ts:490]: since we've gotten here without modifying `entity` and given the first conditional, this is always going to be an empty array right? Using `reduce` in this way seems a little odd to me. Why even use an array as opposed to value/null?

File: src/components/legend.ts
Comment (themadcreator) [inline src/components/legend.ts:514]: and here instead of pushing to `entity`, we just return a statically declared single item array, which seems weird that we've tried to pass around the initial empty array through all the reduce calls.

File: src/components/legend.ts
Comment (themadcreator) [inline src/components/legend.ts]: This seems like a struct, and since you're not exporting it, why not just make it one instead of all the extra getter setter mess?

File: src/components/legend.ts
Comment (themadcreator) [inline src/components/legend.ts:25]: Will "row" and "column" always make sense? Will we ever rotate this text? Will we ever present non-grid-aligned (what about center aligned) legends?

File: src/components/legend.ts
Comment (themadcreator) [inline src/components/legend.ts]: wait, so to calculate the bounds for a column i also have to pass a row?

File: src/components/legend.ts
Comment (themadcreator) [inline src/components/legend.ts]: can we avoid this `indexOf` linear scan if we pass in the row index instead? Will the caller already have that?

File: src/components/legend.ts
Comment (themadcreator) [inline src/components/legend.ts]: we're trading a memory copy for iteration, which is going to be slower in javascript than just scanning the whole array. Still, it seems like the faster thing would just be to do normal for-loop iteration with an early `break`.

Comment (themadcreator) [reply src/components/legend.ts]: Though this is probably premature optimization. Not like I profiled.

File: src/components/legend.ts
Comment (themadcreator) [inline src/components/legend.ts]: `row` is irrelevant here, no? maybe i misunderstood your comment in `.getBounds`

File: test/components/legendTests.ts
Comment (themadcreator) [inline test/components/legendTests.ts:6]: all the other classes here are variables. Should this be?

**Review by themadcreator — CHANGES_REQUESTED** [review]

File: src/components/legend.ts
Comment (hellochar) [reply src/components/legend.ts]: Yeah lets put this in MathUtils

**Review by hellochar — COMMENTED** [review]
Code looks mostly fine.

Looks like the layout is a bit more spread out vertically -- do we know why that is?

| Old  | New |
| ------------- | ------------- |
| image | image |

There's more space between the second and third row.

-----

There's some craziness with how Legend behaves inside a Table. You should be able to control how much space e.g. a plot gets vs the legend by using Table.rowWeight or Table.columnWeight:

No column weights:
image

Plot column weight 3:
image

But it looks like develop has a bug where Legend only respects column weight if the Legend would've been rendered with text ellipses even if the other plots had no weight. If I change the previous screenshots to have a slightly less long label, then we get:

image

Irrespective of Plot column weight. This is where `maxWidth` would be most valuable, but if it was because of a bug to begin with, I'm now a little worried that `maxWidth` may not make sense at all. Thoughts on what to do with maxWidth and this bug?

File: src/components/legend.ts
Comment (CalvinFernandez) [reply src/components/legend.ts]: sure thing

File: src/components/legend.ts
Comment (CalvinFernandez) [reply src/components/legend.ts:25]: hmm good question -- rotated text should still be doable by the table (rotate the table :p ) I think even center-aligned legends are still a subset of a grid, just one with preference for changing margins such that everything is centered by row. Thoughts?

File: src/components/legend.ts
Comment (CalvinFernandez) [reply src/components/legend.ts]: lol yea the API is a little goofy i can change this if you think its too crazy

File: src/components/legend.ts
Comment (CalvinFernandez) [reply src/components/legend.ts]: sure

File: src/components/legend.ts
Comment (CalvinFernandez) [reply src/components/legend.ts]: extra infinity is the default value that will be returned if, say, `this.maxWidth` returns null .

File: src/components/legend.ts
Comment (CalvinFernandez) [reply src/components/legend.ts]: I can just fix that api

File: src/components/legend.ts
Comment (CalvinFernandez) [reply src/components/legend.ts]: ugh i know

File: src/components/legend.ts
Comment (themadcreator) [reply src/components/legend.ts]: Yeah, how about `getRowBounds` and `getColumnBounds`

File: src/components/legend.ts
Comment (themadcreator) [reply src/components/legend.ts]: But you have **two** `Infinity`s, one in the array, one as the second argument.

File: src/components/legend.ts
Comment (CalvinFernandez) [reply src/components/legend.ts:478]: nope just here

**Comment by hellochar** [comment]
Behavior feels much better; lgtm after build failure!

**Review by hellochar — APPROVED** [review]

File: src/components/legend.ts
Comment (themadcreator) [inline src/components/legend.ts:6]: Tight

File: src/components/legend.ts
Comment (themadcreator) [reply src/components/legend.ts:25]: Im still not sure about center alignment, but this is fine.

File: src/components/legend.ts
Comment (themadcreator) [inline src/components/legend.ts:408]: Are we guaranteed that the `length` will always be a multiple of 2?

File: src/components/legend.ts
Comment (themadcreator) [inline src/components/legend.ts:130]: yeah, much nicer

**Review by themadcreator — APPROVED** [review]
Solid

File: src/components/legend.ts
Comment (CalvinFernandez) [reply src/components/legend.ts:408]: yep

**Pull request merged**
